### PR TITLE
[WIP] Adding priors

### PR DESCRIPTION
**Pull request by LukasSp**
Add prior to your objective function

### CODE REVIEW

**Comment by yannikschaelte** [comment]
there still seem to be some merging problems (those lines starting with ``>>>>>>>`` or ``<<<<<<<<``, at least in ``__init__.py``. @LukasSp please check that, then we can have a closer look. but looks good from a first impression!

File: pypesto/__init__.py [hunk 1]
```diff
@@ -38,6 +39,10 @@
            "Optimizer",
            "ScipyOptimizer",
            "DlibOptimizer",
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (paulstapor) [reply pypesto/__init__.py]: This means unmerged changes... please clean that one up.

File: pypesto/__init__.py [hunk 2]
```diff
@@ -38,6 +39,10 @@
...
+           "Prior"]
+=======
            "profile",
            "ProfileOptions",
            "ProfilerResult"]
+>>>>>>> ICB-DCM/master
```
Comment (paulstapor) [inline pypesto/__init__.py]: This is the end of these unmerged changes

File: pypesto/objective/objective.py [hunk 3]
```diff
@@ -101,13 +101,18 @@ def __init__(self,
                  res=None, sres=None,
                  fun_accept_sensi_orders=False,
                  res_accept_sensi_orders=False,
+                 prior = None,
```
Comment (paulstapor) [inline pypesto/objective/objective.py]: That one looks not conformal with pep8
Run ./run_flake8.sh locally to get all formatting issues, fix them and repush ;)

File: pypesto/objective/prior.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            log_index = None
+            logE_index = None
+
+        
+        ##TODO
+        #check if all parameters are already zero
+        #print(x)
+        #if x.all()<1e-4: flag=1
+        #Hessian
+
```
[the remark this answers is not in the document]
Comment (paulstapor) [reply pypesto/objective/prior.py]: Again something that flake8 will complain about

File: pypesto/objective/prior.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        #if x.all()<1e-4: flag=1
+        #Hessian
+        
+        norm_index_1 = [i for i, j in enumerate(self.priorType_list) if j=='norm']
+        lap_index_1  = [i for i, j in enumerate(self.priorType_list) if j=='lap']
+
+        norm_index = np.intersect1d(norm_index_1,estimate)
+        lap_index  = np.intersect1d(lap_index_1,estimate)
+
+        from scipy.stats import multivariate_normal, laplace, norm
```
[the remark this answers is not in the document]
Comment (paulstapor) [reply pypesto/objective/prior.py]: isn't there a norm function in numpy as well?
I would rather go for that one, but it should not make a difference actually...

**Comment by paulstapor** [comment]
Is this pr now at the level of being able to merge? If so, please tell me, and I will go over the code again and possibly approve...

**Comment by LukasSp** [comment]
> Is this pr now at the level of being able to merge? If so, please tell me, and I will go over the code again and possibly approve...

Not yet... I found some improvements and added some code.. I will let you know... Thank you

**Comment by LeonardSchmiester** [comment]
The prior calculation requires some code from an external toolbox for the logicle scale (and a wrapper written by Lukas around it). 

How do we handle this? Put the logicle toolbox as install_requires(?) and add the wrapper from Lukas directly to pyPESTO? Then we could also use the logicle scale as parameter transform.

**Comment by yannikschaelte** [comment]
> The prior calculation requires some code from an external toolbox for the logicle scale (and a wrapper written by Lukas around it).
> 
> How do we handle this? Put the logicle toolbox as install_requires(?) and add the wrapper from Lukas directly to pyPESTO? Then we could also use the logicle scale as parameter transform.

which toolbox?

**Comment by LeonardSchmiester** [comment]
The Code is taken from here: https://github.com/bpteague/cytoflow/tree/master/cytoflow/utility/logicle_ext

**Comment by yannikschaelte** [comment]
what's the plan with this pull request? pretty long-time stale.

**Comment by LeonardSchmiester** [comment]
> 
> 
> what's the plan with this pull request? pretty long-time stale.

Still not 100% sure how to best handle it with the logicle scale from cytoflow.

I also need to check Lukas code in detail as I was not involved in that.

**Comment by yannikschaelte** [comment]
closing as this is being re-implemented all over.

**Pull request closed without merging**